### ISSUE TITLE
Add new tool to support copying/filtering certificates

### ISSUE THREAD

**Issue by atc0005** [atc0005/check-cert#171]
E.g, rough equivalent of:

- `echo -n | openssl s_client -connect $HOST:$PORTNUMBER -servername $SERVERNAME | openssl x509 > /tmp/$SERVERNAME.crt`

~But allow saving either all certs or just specific types of certs:~

- ~leaf~
- ~intermediates~
- ~root~

~or even chain position:~

- ~0~
- ~1~
- ~2~
- ~...~

**EDIT**:

Moved those features to separate GH issues, keeping this one focused just on the initial prototype implementation.

Feedback on this initial implementation (included in the [v0.19.0 release](https://github.com/atc0005/check-cert/milestone/107)) is welcome. Please provide your responses here:

- https://github.com/atc0005/check-cert/discussions/963

refs:

- https://serverfault.com/questions/139728/how-to-download-the-ssl-certificate-from-a-website

**Comment by atc0005** [comment]
Credit: Idea inspired by a conversation with Phil. You know who you are. :)

**Comment by atc0005** [comment]
Additional answers from the serverfault page suggest:

- `openssl s_client -connect $server:$port -servername $server -showcerts 2>&1 < /dev/null | sed -n '/-----BEGIN/,/-----END/p' > chain.pem`
- `gnutls-cli --print-cert myserver.com </dev/null| sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > myserver.crt`

Longer answer:

```shell
sudo apt-get install gnutls-bin 
gnutls-cli --print-cert myserver.com </dev/null| sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > myserver.crt
sudo cp myserver.crt /usr/local/share/ca-certificates/myserver.crt
sudo update-ca-certificates
```

**Comment by atc0005** [comment]
Maybe instead of adding the support to the existing `lscert`, perhaps creating a new tool dedicated to this purpose is a better approach. Maybe `cpcert`?

------

### Additional thoughts

#### Filter support

I could see flags added later to specify which types of certificates to filter in/out of the resulting output file. This same file could even use an existing (local) cert file as an input source, providing the same filtering options.

For example:

- keep everything
  - makes the most sense if using remote source
  - if using a local file, could provide automatic conversion between binary DER input and PEM output
- keep leaf + intermediates
- keep leaf cert only
- keep intermediates only

and potentially other combinations as they make sense.

#### A `fix` flag

This flag could perform automatic fixing of input certificate chain files:

- reordering of cert chain elements
  - this could allow ingesting a cert file with completely reversed elements or any other combination of invalid ordering and output a valid certificate chain
- building a complete chain from an incomplete input file
  - use AIA support (http://www.pkiglobe.org/auth_info_access.html) to allow ingesting a leaf cert and generating a complete chain output file from fetched intermediates associated with that leaf cert

#### TODO

Note to self:

I'd need to spin off separate GH issues for this as not all features would be feasible to provide for the initial implementation.

**Comment by atc0005** [comment]
Feedback on this initial implementation (included in the [v0.19.0 release](https://github.com/atc0005/check-cert/milestone/107)) is welcome. Please provide your responses here:

- https://github.com/atc0005/check-cert/discussions/963
